[클린코드 7기 이상혁] 자동차 경주 미션 STEP 4

sanghyuk-2i opened the pull request:
😀 안녕하세요, 유조님!

4번째 미션을 모두 수행하여 리뷰를 요청드립니다.

요구 사항
- [x] 자동차에 이름을 부여할 수 있다. 전진하는 자동차를 출력할 때 자동차 이름을 같이 출력한다.
- [x] 자동차 이름은 쉼표(,)를 기준으로 구분하며 이름은 5자 이하만 가능하다.
- [x] 자동차 경주는 5회로 고정하여 진행한다.
- [x] 자동차는 1회당 1칸씩 전진한다
- [x] 회차를 거듭할 때마다 자동차가 지나간 궤적을 출력한다(실행 예시 참고).
- [x] 사용자가 잘못된 입력 값을 작성한 경우 프로그램을 종료한다.
- [x] **전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.**
- [x] **자동차 경주 게임을 완료한 후 누가 우승했는지를 알려준다. 우승자는 한 명 이상일 수 있다.**
- [x] **게임의 시도 횟수를 직접 입력 받을 수 있다.**

sanghyuk-2i committed:
fix: step 3 피드백 반영

- 같은 context를 그룹화하여 테스트 명세서 보완
- 테스트 코드 맥락 보완
- main 관련 테스트가 아닌 utils 함수로 파일 이름 변경
- isInputValid 버그 수정 및 리팩토링

sanghyuk-2i committed:
feat: 게임의 시도 횟수를 입력 가능 추가

- 게임 시도 횟수를 입력하는 유효성 검사 추가 및 테스트 코드 작성
- 네이밍 변경
- step 4 요구사항 추가

sanghyuk-2i committed:
feat: 우승자 기능 추가

- 자동차 경주 게임을 완료한 후 누가 우승했는지를 알려준다. 우승자는 한 명 이상일 수 있다.
- 관련된 테스트 코드 작성

sanghyuk-2i committed:
feat: 전진하는 조건 기능 추가

- 전진하는 조건 정책 추가 및 관련된 테스트 코드 작성
- Random 테스트 코드 작성

yujo11 commented on __tests__/utils/isCarNamesValid.test.js:
```diff
@@ -0,0 +1,57 @@
+import { isCarNamesValid } from '../../src/utils/index.js';
+
+describe('isCarNamesValid 함수 테스트', () => {
+  describe('쉼표를 어떻게 사용하는냐에 따라 자동차 이름을 구분한다.', () => {
```
계층을 나눠주셨군요 ㅎㅎ 좋네용 👍 👍

yujo11 commented on __tests__/utils/isCarNamesValid.test.js:
```diff
@@ -0,0 +1,57 @@
+    });
+  });
+
+  describe('5글자 이하의 문자열만 이름으로 사용할 수 있어야 한다.', () => {
```
`5글자`가 기준이니 경계값인 4글자, 6글자의 이름을 테스트 해봐도 좋을거 같아요 ㅎㅎ

yujo11 commented on __tests__/utils/isCarNamesValid.test.js:
```diff
@@ -0,0 +1,57 @@
+
+describe('isCarNamesValid 함수 테스트', () => {
+  describe('쉼표를 어떻게 사용하는냐에 따라 자동차 이름을 구분한다.', () => {
+    test('"아반떼,그렌져"을 입력하면 올바르게 처리해야 한다.', () => {
+      const isValid = isCarNamesValid('아반떼,그렌져');
+
+      expect(isValid).toBeTruthy();
+    });
+
+    test('"아반떼, 그렌져"을 입력하면 올바르게 처리해야 한다.', () => {
+      const isValid = isCarNamesValid('아반떼, 그렌져');
+
+      expect(isValid).toBeTruthy();
+    });
+
+    test('"아반떼,"을 입력하면 올바르게 처리할 수 없게 한다.', () => {
```
"올바르게 처리한다", "올바르게 처리할 수 없게 한다" 라는게 어떤 결과 값을 기대하는 테스트인지 직관적으로 다가오지 않는데요. validation을 수행하는 함수에 대한 테스트이니 "true를 반환한다." "false를 반환한다." 처럼 적어도 좋을거 같다는 생각이 드네용!

yujo11 commented on src/utils/index.js:
```diff
@@ -1,21 +1,45 @@
+  if (!input) return false;
 
-  return hasSplitedValues && hasNonEmptyValues;
+  const splitedValues = input.split(',');
```
밑에 로직을 보면 `trim()`이 반복해서 들어가고 있는데 여기에서 `trim()`을 미리 해줘도 좋을거 같네요
```suggestion
  const splitedValues = input.split(',').map((it) => it.trim());
```

yujo11 commented on src/services/Game.js:
```diff
@@ -73,6 +77,22 @@ export default class Game {
+   * @param {Car[]} cars 마지막 바퀴까지 달린 자동차들
+   * @returns 우승한 자동차 이름들
+   */
+  getWinners(cars) {
+    const carLocations = cars.map((car) => car.getLocation());
+    const topTrack = Math.max(...carLocations);
+
+    return cars
+      .filter((car) => topTrack === car.getLocation())
+      .map((car) => car.getName())
+      .join(', ');
+  }
```
우승한 자동차들의 "이름"을 반환하고 있다는 것이 함수명을 통해 더 잘 드러났으면 좋겠네용
```suggestion
  getWinnersName(cars) {
    const carLocations = cars.map((car) => car.getLocation());
    const topTrack = Math.max(...carLocations);

    return cars
      .filter((car) => topTrack === car.getLocation())
      .map((car) => car.getName())
      .join(', ');
  }
```

yujo11 commented on src/utils/index.js:
```diff
@@ -1,21 +1,45 @@
+ * @param {number} max 난수의 최대 범위의 값
+ * @returns 랜덤 값
+ */
+export function getRandomNumber(min = 0, max = 10000) {
```
`min`, `max`의 default parameter가 0, 10000으로 설정 된 이유가 있을까용?

sanghyuk-2i replied:
사실, 확장성을 생각해서 아무 생각없이 10_000으로 default 값 처리한 것 같습니다.
일단, 요구사항의 기준대로 처리하고, 향후 변경되면 그때 다시 고려하겠습니다.

```js
export function getRandomNumber(min = 0, max = 9) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}
```

sanghyuk-2i replied:
사실, 확장성을 생각해서 아무 생각없이 10_000으로 default 값 처리한 것 같습니다.
일단, 요구사항의 기준대로 처리하고, 향후 변경되면 그때 다시 고려하겠습니다.

```js
export function getRandomNumber(min = 0, max = 9) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}
```

yujo11 commented on src/utils/index.js:
```diff
@@ -1,21 +1,45 @@
+ * @param {number} max 난수의 최대 범위의 값
+ * @returns 랜덤 값
+ */
+export function getRandomNumber(min = 0, max = 10000) {
```
소소하게 1,000 단위를 넘어가는 숫자들에는 numeric separators를 같이 사용해주시면 좋을거 같네용 ㅎㅎ

```suggestion
export function getRandomNumber(min = 0, max = 10_000) {
```

yujo11 commented on src/utils/index.js:
보통 util 함수라고 부르는 친구들은 도메인 로직과 무관하게 재사용 가능한 함수들인데요.

현재 `utils/`에 있는 함수들은 `getRandomNumber`를 제외하고는 도메인과 얽혀있는 부분들이 많이 보이는거 같아요. 함수의 위치에 대해서도 한번 고민해보시면 좋을거 같습니다 !

sanghyuk-2i replied:
폴더 구조를 변경해서 해당 내용 개선해 봤습니다!

yujo11 commented on src/services/Game.js:
```diff
@@ -82,9 +102,13 @@ export default class Game {
    */
   getPlayerCurrentStatus(car) {
```
`getPlayerCurrentStatus`라는 이름만 보면 자동차의 "현재" 상태를 반환하는 것처럼 느껴지는데 실제로는 이 함수가 호출 되는 시점에 자동차가 한 번 더 움직일 수 있다는 점이 조금 혼란스럽게 느껴지네용 🤔

어떻게 개선해 볼 수 있을까요?

sanghyuk-2i replied:
단순하게 이름을 변경하려다가, 함수 안에 처리할 부분들이 많아서 저도 어떻게 이름을 변경해야 할지 어려웠습니다.
그래서, 함수 안에 로직들을 좀 더 쪼개니까 수행하는 역할대로 이름을 쉽게 지을 수 있었던 것 같습니다.
움직임 결정을 짓는 함수를 중심으로 '움직여야 할 때 수행할 함수' 또는 '움직일 필요가 없을 때 수행할 함수'를 실행할 수 있도록 
변경했습니다.

이렇게 개선하면 괜찮을지 궁금합니다.

```js
  /**
   * 움직임이 필요할 때, 실행되는 Callback 함수
   *
   * @param {Car} car 자동차 정보
   * @returns 자동차 이름과 자동차의 위치 정보
   */
  handleCarMove(car) {
    car.moveForward();

    return { name: car.getName(), location: car.getLocation() };
  }

  /**
   * 움직임이 필요가 없을 때, 실행되는 Callback 함수
   *
   * @param {Car} car 자동차 정보
   * @returns 자동차 이름과 자동차의 위치 정보
   */
  handleCarStay(car) {
    return { name: car.getName(), location: car.getLocation() };
  }

  /**
   * 자동차의 움직임 여부을 결정하는 함수
   *
   * @param {Car} car 자동차 정보
   * @param {function(Object): void} onMove 움직일 필요가 있어 작동되는 callback
   * @param {function(Object): void} onStay 움직일 필요가 없어 작동되는 callback
   * @returns 자동차 이름과 자동차의 위치 정보
   */
  determineCarMovement(car, onMove, onStay) {
    const randomNumber = getRandomNumber();
    if (randomNumber < MIN_MOVEMENT_THRESHOLD) {
      return onStay(car);
    }

    return onMove(car);
  }
```

yujo11 replied:
오 적절한 개선 방향인거 같습니다! 좋네요!! 👍 👍 👍

yujo11 reviewed (changes requested):
상혁님 안녕하세요 ~~

지난 step에 남긴 코멘트들도 잘 확인하고 반영해주셨군요 감사합니다 ㅎㅎ 👍 👍 👍 

이번에도 코멘트들 몇 가지 남겼는데요. 이 PR에서 확인하고 넘어가면 좋을거 같아서 request changes로 드립니다!

코멘트 확인해보시고 다시 리뷰 요청 부탁드립니다! 고생 많으셨어용 ~~ 👍 👍 👍

sanghyuk-2i committed:
fix: 피드백 내용 반영

- 테스트 코드를 좀 더 명확하게 수정
- 자동차 클래스의 함수명 변경
- 폴더 구조 개선으로 도메인 관련 파일 이동
- getRandomNumber 기본 parameter 값 변경

sanghyuk-2i commented:
좋은 리뷰 내용들 너무 감사합니다 🙏
말씀해 주신 내용들을 최대한 반영을 했으며, 변경이 큰 부분들을 남겨주신 코멘트의 댓글로 수정 내용을 작성해 봤습니다.

yujo11 commented on src/domains/game/utils.js:
👍 👍 👍

yujo11 reviewed (approved):
리팩터링까지 고생 많으셨습니다 상혁님 !!

리뷰 반영해주실 때도 하나의 커밋으로 모두 남기는게 아니라 하나의 변경마다 하나의 커밋으로 남으면 좋을거 같네요!

이제 마지막 step만 남았네요! 다음 step에서 뵙겠습니다 ~~ 👍 👍 👍
